contracts/kernel-gateway/KernelCoGateway.sol

```diff
@@ -0,0 +1,196 @@
+pragma solidity >=0.5.0 <0.6.0;
+
+// Copyright 2019 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "../proxies/MasterCopyNonUpgradable.sol";
+import "../message-bus/MessageInbox.sol";
+import "../consensus/ConsensusModule1.sol";
+import "./KernelBase.sol";
+
+contract KernelCoGateway is
+    MasterCopyNonUpgradable, // Always keep this always at location 0
+    MessageInbox,
+    ConsensusModule1, // TODO: replace this with ConsensusModule
+    KernelBase
+{
+    /* Events */
+
+    event ConfirmedOpenKernelGateway(bytes32 kernelMessageHash);
+
+
+    /* Constants */
+
+    /** This is the storage index of MessageInbox::inbox mapping. */
+    uint8 constant public INBOX_OFFSET = 1;
+
+
+    /* External Functions */
+
+    /**
+     * @notice Setup kernel cogateway contract.
+     * @param _chainId Chain identifier.
+     * @param _kernelGateway KernelGateway contract address.
+     * @param _outboxStorageIndex Storage index of outbox mapping in
+     *                            KernelGateway contract.
+     * @param _stateRootProvider State root provider contract address.
+     * @param _maxStorageRootItems Defines how many storage roots should be
+     *                             stored in circular buffer.
+     */
+    function setup(
+        bytes20 _chainId,
+        address _kernelGateway,
+        uint8 _outboxStorageIndex,
+        StateRootI _stateRootProvider,
```

Comment:
Is this the address of `ProtoCore`. Will ProtoCore change when Core on origin is halted or corrupted?